specs/encryption.md

```diff
@@ -0,0 +1,83 @@
+# Encryption format
+
+<dl>
+	<dt>Namespace</dt><dd>`encryption`</dd>
+	<dt>Version</dt><dd>1.0</dd>
+</dl>
+
+TODO: Namespace name is a bit odd - ideas on better one?
+
+This is an application format for describing how to decrypt other elements of the same share.
+
+It's intended that this be handled automatically by the Earthstar libraries, so applications don't need to thing about the details here.
+
+i.e.
+- App starts by writing out some keys under /encryption
+- Then any keys written after are transparently encrypted (both path and/or payload) before being stored
+- And likewise any list/get of the keyspace happens after decryption
+
+This also lends itself to different users of the same application format having different setups for encryption, and other clients reading it will apply the correct settings.
+
+## Motivation
+
+Trying to predict everything about how different users will want to use encryption of paths and payloads is challenging, to say the least.
+
+Decoupling this and allowing this to vary across different clients is helpful.
+
+This also allows for a lot of the detail to be done by the standard library, instead of needing to be coded by individual clients.
+
+### Path Encryption settings
+
+The path format of a path settings document is:
+
+```
+/encryption/1.0/{PATH}/path.yaml
+```
+
+#### fields
+
+```
+rules:
+- from: <utc timestamp> # optional
+  to: <utc timestamp> # optional
+
+  key: self | /keys/key-id # See [the keys spec](keys.md)
+  recursive: true # Whether or not this applies only to one level, or all the way down the tree
+  type: none | path-based | per-key | static
+```
+
+Rules are processed in order, first match applies. (Though making them all disjoint is expected)
+
+##### per-key
+
+This uses the following construction:
+- plaintext path: `<input-public-key>` - this should be the ed25519 public key of the other party
+- encrypted path: `hkdf(key=scalarmult(<subspace-private-key>, <input-public-key>), info=<namespace>#<subspace>#<path>)`
+
+For someone trying to decrypt they can use the opposite construction to find the path encrypted with their key: `hkdf(key=scalarmult(<their-private-key>, <subspace-public-key>), info=<namespace>#<subspace>#<path>)`
+
+##### TODO more details
+
+### Payload Encryption settings
+
+The path format of a payload settings document is:
+
+```
+/encryption/1.0/{PATH}/payload.yaml
```

Comment:
Was hacking on this here: https://github.com/earthstar-project/earthstar/pull/328
Ended up deciding that embedding the path was painful - now it's just "load all the documents in, the docs have rules in, take most specific rule. Easier to implement"
So this bit needs rewriting